File: android/src/main/kotlin/com/securebankkit/flutter_security_suite/SecureBankKitPlugin.kt
package com.securebankkit.flutter_security_suite

import android.app.Activity
import android.content.Context
import androidx.annotation.NonNull
import com.securebankkit.flutter_security_suite.handlers.AppIntegrityHandler
import com.securebankkit.flutter_security_suite.handlers.RootDetectionHandler
import com.securebankkit.flutter_security_suite.handlers.ScreenshotHandler
import com.securebankkit.flutter_security_suite.handlers.SecureStorageHandler
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class SecureBankKitPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var activity: Activity? = null

    private val rootHandler = RootDetectionHandler()
    private val screenshotHandler = ScreenshotHandler()
    private val integrityHandler = AppIntegrityHandler()
    private val storageHandler = SecureStorageHandler()

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "com.securebankkit/security")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            // ── Root Detection ──────────────────────────
            "root#isDeviceRooted" -> {
                result.success(rootHandler.isDeviceRooted())
            }

            // ── Screenshot Protection ───────────────────
            "screenshot#enable" -> {
                activity?.runOnUiThread {
                    screenshotHandler.enable(activity)
                    result.success(null)
                } ?: result.error("NO_ACTIVITY", "No activity attached", null)
            }
            "screenshot#disable" -> {
                activity?.runOnUiThread {
                    screenshotHandler.disable(activity)
                    result.success(null)
                } ?: result.error("NO_ACTIVITY", "No activity attached", null)
            }

            // ── App Integrity ───────────────────────────
            "integrity#isValid" -> {
                result.success(integrityHandler.isAppIntegrityValid(context))
            }

            // ── Secure Storage ──────────────────────────
            "storage#write" -> {
                val key = call.argument<String>("key")
                val value = call.argument<String>("value")
                if (key != null && value != null) {
                    storageHandler.write(context, key, value)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "key and value are required", null)
                }
            }
            "storage#read" -> {
                val key = call.argument<String>("key")
                if (key != null) {
                    result.success(storageHandler.read(context, key))
                } else {
                    result.error("INVALID_ARGS", "key is required", null)
                }
            }
            "storage#delete" -> {
                val key = call.argument<String>("key")
                if (key != null) {
                    storageHandler.delete(context, key)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "key is required", null)
                }
            }
            "storage#deleteAll" -> {
                storageHandler.deleteAll(context)
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    // ── ActivityAware ───────────────────────────────────────

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}

File: ios/Classes/SecureBankKitPlugin.swift
import Flutter
import UIKit

public class SecureBankKitPlugin: NSObject, FlutterPlugin {

    private let jailbreakHandler = JailbreakDetectionHandler()
    private let screenshotHandler = ScreenshotHandler()
    private let integrityHandler = AppIntegrityHandler()
    private let storageHandler = SecureStorageHandler()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.securebankkit/security",
            binaryMessenger: registrar.messenger()
        )
        let instance = SecureBankKitPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {

        // ── Root / Jailbreak Detection ──────────────
        case "root#isDeviceRooted":
            result(jailbreakHandler.isDeviceJailbroken())

        // ── Screenshot Protection ───────────────────
        case "screenshot#enable":
            screenshotHandler.enable()
            result(nil)

        case "screenshot#disable":
            screenshotHandler.disable()
            result(nil)

        // ── App Integrity ───────────────────────────
        case "integrity#isValid":
            result(integrityHandler.isAppIntegrityValid())

        // ── Secure Storage ──────────────────────────
        case "storage#write":
            guard let args = call.arguments as? [String: Any],
                  let key = args["key"] as? String,
                  let value = args["value"] as? String else {
                result(FlutterError(
                    code: "INVALID_ARGS",
                    message: "key and value are required",
                    details: nil
                ))
                return
            }
            let success = storageHandler.write(key: key, value: value)
            if success {
                result(nil)
            } else {
                result(FlutterError(
                    code: "STORAGE_ERROR",
                    message: "Failed to write to keychain",
                    details: nil
                ))
            }

        case "storage#read":
            guard let args = call.arguments as? [String: Any],
                  let key = args["key"] as? String else {
                result(FlutterError(
                    code: "INVALID_ARGS",
                    message: "key is required",
                    details: nil
                ))
                return
            }
            result(storageHandler.read(key: key))

        case "storage#delete":
            guard let args = call.arguments as? [String: Any],
                  let key = args["key"] as? String else {
                result(FlutterError(
                    code: "INVALID_ARGS",
                    message: "key is required",
                    details: nil
                ))
                return
            }
            let _ = storageHandler.delete(key: key)
            result(nil)

        case "storage#deleteAll":
            let _ = storageHandler.deleteAll()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
